src/library_promise.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * @license
+ * Copyright 2023 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+mergeInto(LibraryManager.library, {
+  $promiseMap__deps: ['$handleAllocator'],
+  $promiseMap: "new handleAllocator();",
+
+  $getPromise__deps: ['$promiseMap'],
+  $getPromise: function(id) {
+    return promiseMap.get(id).promise;
+  },
+  emscripten_promise_create__deps: ['$promiseMap'],
+  emscripten_promise_create__sig: 'p',
+  emscripten_promise_create: function() {
+    var promiseInfo = {};
+    promiseInfo.promise = new Promise((resolve, reject) => {
+      promiseInfo.reject = reject;
+      promiseInfo.resolve = resolve;
+    });
+    var id = promiseMap.allocate(promiseInfo);
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_create: ' + id);
+#endif
+    return id;
+  },
+
+  emscripten_promise_destroy__deps: ['$promiseMap'],
+  emscripten_promise_destroy__sig: 'vp',
+  emscripten_promise_destroy: function(id) {
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_destroy: ' + id);
+#endif
+    promiseMap.free(id);
+  },
+
+  emscripten_promise_resolve__deps: ['$promiseMap',
+                                     '$getPromise',
+                                     'emscripten_promise_destroy'],
+  emscripten_promise_resolve__sig: 'vpip',
+  emscripten_promise_resolve: function(id, result, value) {
+#if RUNTIME_DEBUG
+    err('emscripten_promise_resolve: ' + id);
+#endif
+    var info = promiseMap.get(id);
+    if (result == {{{ cDefine('EM_PROMISE_FULFILL') }}}) {
+      info.resolve(value);
+    } else if (result == {{{ cDefine('EM_PROMISE_MATCH') }}}) {
+      info.resolve(getPromise(value));
+    } else if (result == {{{ cDefine('EM_PROMISE_MATCH_RELEASE') }}}) {
+      info.resolve(getPromise(value));
+      _emscripten_promise_destroy(value);
+    } else if (result == {{{ cDefine('EM_PROMISE_REJECT') }}}) {
+      info.reject(value);
+    } else {
+      abort('invalid promise result');
+    }
+  },
+
+  $makePromiseCallback__deps: ['$getPromise',
+                               '$POINTER_SIZE',
+                               'emscripten_promise_destroy'],
+  $makePromiseCallback: function(callback, userData) {
+    return (value) => {
+#if RUNTIME_DEBUG
+      dbg('emscripten promise callback: ' + value);
+#endif
+      var stack = stackSave();
+      // Allocate space for the result value and initialize it to NULL.
+      var resultPtr = stackAlloc(POINTER_SIZE);
+      {{{ makeSetValue('resultPtr', 0, '0', '*') }}};
+      try {
+        var result =
+            {{{ makeDynCall('ippp', 'callback') }}}(resultPtr, userData, value);
+        var resultVal = {{{ makeGetValue('resultPtr', 0, '*') }}};
+      } catch (e) {
+        // If the thrown value is potentially a valid pointer, use it as the
+        // rejection reason. Otherwise use a null pointer as the reason.
+#if MEMORY64
+        if (typeof e !== 'bigint') {
+          throw 0n;
+        }
+#else
+        if (typeof e !== 'number') {
+          throw 0;
+        }
```

Comment:
I'm confused by this.. what if a JS exceptions is thrown.. don't we want to re-throw it here?

Reply:
Any value thrown from here will become the "reason" value for the rejection of the new promise returned from `emscripten_promise_then`. If we are in wasm64 and the reason value is not a bigint, we'll get a TypeError when it is converted to a pointer to pass to the next rejection handler. To avoid that type error, we have to avoid throwing anything except a valid pointer here.